### DISCUSSION TITLE
Ratio_arc in RandomBN()

### DISCUSSION THREAD

**Discussion by kenneth-lee-ch** [agrumery/aGrUM#18] [Q&A]
I give a ratio_arc = 3 and set the number of nodes to be 8, how come the resulting graph has only 9 arcs in total? It seems like it is inconsistent with what the documentation describes that `ratio_arc (float) – number of arcs = n * ratio_arc`.

**Comment by phwuil** [comment]
Hello, we use the approach from [Ide & Cozman, 2002]. However, iterating in a MCMC is not done correctly : in our implementation, when the action of adding an arc is called, it may not take place if the constraints are not respected (acyclicity, etc.). It's not right to do this, as it's possible that you won't be able to iterate enough to obtain the required number of arcs.

In your case, you want a graph with 8 nodes and 24 arcs, which isn't very far from the complete graph (26 arcs). It becomes difficult to randomly find syntactically correct arcs... I guess that's the reason.

In the context of a "classic" Bayesian network (ratio>1,<2 I'd say), the algorithm should work fine.

**Reply by phwuil** [reply]
but we need to improve this code  asap :-(

**Comment by kenneth-lee-ch** [comment]
Glad to hear you will fix this asap...

**Comment by kenneth-lee-ch** [comment]
@phwuil One more question, how come I cannot set `arc_ratio` to `0.5`? 

![Screenshot 2023-08-23 at 10 29 02 AM](https://github.com/agrumery/aGrUM/assets/24616330/bf5beb8c-904c-4848-9f32-2ee01c6d38a7)

**Comment by phwuil** [comment]
We want to generate connected graphs. The minimal connected graph is a tree. A tree of n nodes has n-1 arcs. So the ratio 0.5 is not possible (except if n=2 which is not authorized neither)...

**Reply by phwuil** [reply]
![image](https://github.com/agrumery/aGrUM/assets/5980636/8e4ca4a4-6bee-4d01-968c-892174d0e846)
